### ISSUE TITLE
Validation for "create new item"

### ISSUE THREAD

**Issue by guettli** [yourlabs/django-autocomplete-light#652]
In my case I want some basic validation before the user can do "create new item".

Use case: The user enters mail-adresses. I don't to accept "foo@example", but I want to accept "foo@example.com".

I looked at the source. I guess validation is not possible at the moment.

What do you think?

Here is the relevant code:

```

    def post(self, request):
        """Create an object given a text after checking permissions."""
        if not self.has_add_permission(request):
            return http.HttpResponseForbidden()

        if not self.create_field:
            raise ImproperlyConfigured()

        text = request.POST.get('text', None)

        if text is None:
            return http.HttpResponseBadRequest()

        result = self.create_object(text)

        return http.HttpResponse(json.dumps({
            'id': result.pk,
            'text': six.text_type(result),
        }))

```

**Comment by jpic** [comment]
Your are correct, there is no form nor validation support in the create
option.

For complete validation support, I​ recommend you try django-addanother. It
uses a popup like django admin to support a form.

**Comment by guettli** [comment]
I know that django-addanother exists. 

In this case I think "Create new" could be extended, since I only supply a simple string. I don't need a form or popup.

Is there a way to stop "Create new" from appearing via a method on the Select2QuerySetView (one if its mixins)?

If this is possible, this could be solved in a very re-usable way: the developer can supply a check_input() method. If it returns True "Create new" appears, if it returns False, it does not appear.

**Comment by jpic** [comment]
I think I understand now, excellent idea ! 

Perhaps we should also [run this check](https://github.com/yourlabs/django-autocomplete-light/blob/master/src/dal/views.py#L75) ?

**Comment by guettli** [comment]
Maybe has_add_permission() does work. But this name would confuse everyone who has not read this discussion. Since this is not about "permission"

I would suggest a new method: input_can_be_added(self, text)

**Comment by guettli** [comment]
BTW the string which as entered by the users has several names in the library. Once it is "text" once it is "q". Would be nice to agree on one name :-)

**Comment by jpic** [comment]
I meant we should check has_add_permission *as well as* input_can_be_added.

Perhaps can_create_option(text) would be easier to type

**Comment by guettli** [comment]
... nice, you understand me. 

What are we adding here? In the docs I found this:

> Creation of new choices in the autocomplete form

So I would suggest to call the method `can_create_new_choice`.

Although I prefer "valid" or "validation" to be in it.

... it is difficult to name things properly.

**Comment by jpic** [comment]
Perhaps validate_input() ?

**Comment by guettli** [comment]
Now we have three names: q, text, input ....

Sorry, I have no matching name. Maybe tomorrow.

**Comment by jpic** [comment]
If the method that creates the object is create_object(), should it be just can_create_object() ?

**Comment by guettli** [comment]
I am unsure wether the widget is the right place to handle this.

The widget exists to display data.

Adding might be more a task for form.Field.

**Comment by jpic** [comment]
The widget here doesn't create any data ! It just provides a way to create it yourself through a POST request, and the object is created by the view, like with add-another ;)

In 3.0, such values were created in form.Field, but Django doesn't support this and that caused numerous bugs: https://github.com/yourlabs/django-autocomplete-light/issues/604

**Comment by guettli** [comment]
ok, you have more experience in that field as I have. My current solution is to overwrite the whole `post()` method.

Before a matching method name can be found, I think the fundamental question should be solved:

How to call the string the user send to the widget:

 * q
 * text
 * input
 *  ....

**Comment by jpic** [comment]
Seems like it would be better to choose from one that we already have, I think q makes more sense here because text is pretty generic. Would that work for you ?

**Comment by guettli** [comment]
I don't like one-character variables. And in django-context there is already Q.

But I think `q` is the best choice.

**Comment by kevin-brown** [comment]
I'm going to second the idea that "q" is an unfortunate choice for a variable name.

> How to call the string the user send to the widget

Note that on the Select2 side this is usually referred to as the "search term" (both `q` and `term` are sent when searching).

**Comment by jpic** [comment]
term also works for me !​

**Comment by guettli** [comment]
Yes, `term` sound good.

**Comment by jpic** [comment]
Agreed ! Hey are you going to europycon ? Perhaps we could have a hangouts session and talk this through ?

**Comment by guettli** [comment]
@jpic  sorry for the late response. No, I won't go to europycon. I have wife and family, my spare time is limited. I would like to go there :-)

**Comment by jpic** [comment]
Well it was last month ! Anyway, would you like to have a hangouts session
next month or are you all set ?

**Comment by guettli** [comment]
@jpic Yes, I  have some time. What are you interested in? You can reach me at guettliml@thomas-guettler.de

**Comment by lwm** [comment]
Duplicate of https://github.com/yourlabs/django-autocomplete-light/issues/779?
